scene/3d/spring_bone_collision_3d.cpp

```diff
@@ -0,0 +1,189 @@
+			WARN_PRINT("Bone index out of range! Cannot connect BoneAttachment to node!");
+			bone = -1;
+		} else {
+			bone_name = sk->get_bone_name(bone);
+		}
+	}
+
+	notify_property_list_changed();
+}
+
+int SpringBoneCollision3D::get_bone() const {
+	return bone;
+}
+
+void SpringBoneCollision3D::set_position_offset(const Vector3 &p_offset) {
+	if (position_offset == p_offset) {
+		return;
+	}
+	position_offset = p_offset;
+	sync_pose();
+#ifdef TOOLS_ENABLED
+	update_gizmos();
+#endif // TOOLS_ENABLED
+}
+
+Vector3 SpringBoneCollision3D::get_position_offset() const {
+	return position_offset;
+}
+
+void SpringBoneCollision3D::set_rotation_offset(const Quaternion &p_offset) {
+	if (rotation_offset == p_offset) {
+		return;
+	}
+	rotation_offset = p_offset;
+	sync_pose();
+#ifdef TOOLS_ENABLED
+	update_gizmos();
+#endif // TOOLS_ENABLED
+}
+
+Quaternion SpringBoneCollision3D::get_rotation_offset() const {
+	return rotation_offset;
+}
+
+void SpringBoneCollision3D::sync_pose() {
+	if (bone >= 0) {
+		Skeleton3D *sk = get_skeleton();
+		if (sk) {
+			Transform3D tr = sk->get_global_transform() * sk->get_bone_global_pose(bone);
+			tr.origin += tr.basis.get_rotation_quaternion().xform(position_offset);
+			tr.basis *= Basis(rotation_offset);
+			set_global_transform(tr);
+		}
+	}
+}
+
+Transform3D SpringBoneCollision3D::get_transform_from_skeleton(const Transform3D &p_center) const {
+	Transform3D gtr = get_global_transform();
+	Skeleton3D *sk = get_skeleton();
+	if (sk) {
+		Transform3D tr = sk->get_global_transform();
+		gtr = tr.affine_inverse() * p_center * gtr;
+	}
+	return gtr;
+}
+
+void SpringBoneCollision3D::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("get_skeleton"), &SpringBoneCollision3D::get_skeleton);
+
+	ClassDB::bind_method(D_METHOD("set_bone_name", "bone_name"), &SpringBoneCollision3D::set_bone_name);
+	ClassDB::bind_method(D_METHOD("get_bone_name"), &SpringBoneCollision3D::get_bone_name);
+
+	ClassDB::bind_method(D_METHOD("set_bone", "bone"), &SpringBoneCollision3D::set_bone);
+	ClassDB::bind_method(D_METHOD("get_bone"), &SpringBoneCollision3D::get_bone);
+
+	ClassDB::bind_method(D_METHOD("set_position_offset", "offset"), &SpringBoneCollision3D::set_position_offset);
+	ClassDB::bind_method(D_METHOD("get_position_offset"), &SpringBoneCollision3D::get_position_offset);
+
+	ClassDB::bind_method(D_METHOD("set_rotation_offset", "offset"), &SpringBoneCollision3D::set_rotation_offset);
+	ClassDB::bind_method(D_METHOD("get_rotation_offset"), &SpringBoneCollision3D::get_rotation_offset);
+
+	ADD_PROPERTY(PropertyInfo(Variant::STRING_NAME, "bone_name"), "set_bone_name", "get_bone_name");
+	ADD_PROPERTY(PropertyInfo(Variant::INT, "bone", PROPERTY_HINT_NONE, "", PROPERTY_USAGE_NO_EDITOR), "set_bone", "get_bone");
+
+	ADD_PROPERTY(PropertyInfo(Variant::VECTOR3, "offset/position"), "set_position_offset", "get_position_offset");
+	ADD_PROPERTY(PropertyInfo(Variant::QUATERNION, "offset/rotation"), "set_rotation_offset", "get_rotation_offset");
```

Comment:
Properties with a slash are pretty awkward to use from code, and shouldn't be used. We might have a few odd ones out still in the API from 3.x times, or some generated ones (with an ID suffix), but I don't think new properties should make the same mistake.

```suggestion
	ADD_PROPERTY(PropertyInfo(Variant::VECTOR3, "position_offset"), "set_position_offset", "get_position_offset");
	ADD_PROPERTY(PropertyInfo(Variant::QUATERNION, "rotation_offset"), "set_rotation_offset", "get_rotation_offset");
```